In [2]:
import pandas as pd

# Load CSV files into DataFrames
df_lang = pd.read_csv('/content/HCK_HEC_LANG.csv')
df_skills = pd.read_csv('/content/HCK_HEC_SKILLS.csv')
df_staffing = pd.read_csv('/content/HCK_HEC_STAFFING.csv')
df_user = pd.read_csv('/content/HCK_HEC_USER.csv')
df_xp = pd.read_csv('/content/HCK_HEC_XP.csv')

# Display the first few rows of each DataFrame
print("HCK_HEC_LANG:")
print(df_lang.head())

print("\nHCK_HEC_SKILLS:")
print(df_skills.head())

print("\nHCK_HEC_STAFFING:")
print(df_staffing.head())

print("\nHCK_HEC_USER:")
print(df_user.head())

print("\nHCK_HEC_XP:")
print(df_xp.head())


HCK_HEC_LANG:
   USER_ID LANGUAGE_SKILL_DSC  LANGUAGE_SKILL_LVL
0  2843838            English               100.0
1  2843838             French               100.0
2  2841644            English                90.0
3  2818577             French               100.0
4  2452725             French               100.0

HCK_HEC_SKILLS:
   USER_ID            SKILLS_DSC     DOMAIN_DSC  LEVEL_VAL
0  2836571  Snowflake Data Cloud  Data Platform       60.0
1  2836571       Microsoft Azure  Data Platform       60.0
2  2836571  Snowflake Data Cloud  Data Platform       60.0
3  2836571       Microsoft Azure  Data Platform       60.0
4  2836571  Snowflake Data Cloud  Data Platform       60.0

HCK_HEC_STAFFING:
   USER_ID  MONTH_1  MONTH_2  MONTH_3  MONTH_4  MONTH_5  MONTH_6  MONTH_7  \
0  2843838      100      100      100      100      100      100      100   
1  2479537      100      100      100      100      100      100      100   
2  2533337       25       25       25       25       25       25 

In [3]:
# Get column names for each DataFrame
print("HCK_HEC_LANG Columns:")
print(df_lang.columns.tolist())

print("\nHCK_HEC_SKILLS Columns:")
print(df_skills.columns.tolist())

print("\nHCK_HEC_STAFFING Columns:")
print(df_staffing.columns.tolist())

print("\nHCK_HEC_USER Columns:")
print(df_user.columns.tolist())

print("\nHCK_HEC_XP Columns:")
print(df_xp.columns.tolist())


HCK_HEC_LANG Columns:
['USER_ID', 'LANGUAGE_SKILL_DSC', 'LANGUAGE_SKILL_LVL']

HCK_HEC_SKILLS Columns:
['USER_ID', 'SKILLS_DSC', 'DOMAIN_DSC', 'LEVEL_VAL']

HCK_HEC_STAFFING Columns:
['USER_ID', 'MONTH_1', 'MONTH_2', 'MONTH_3', 'MONTH_4', 'MONTH_5', 'MONTH_6', 'MONTH_7', 'MONTH_8', 'MONTH_9', 'MONTH_10', 'MONTH_11', 'MONTH_12']

HCK_HEC_USER Columns:
['USER_ID', 'LAST_NAME', 'FIRST_NAME', 'ANNEES_XP']

HCK_HEC_XP Columns:
['USER_ID', 'MISSION_DSC']


In [4]:
# Calculate the number of unique USER_IDs in each table
unique_user_lang = df_lang['USER_ID'].nunique()
unique_user_skills = df_skills['USER_ID'].nunique()
unique_user_staffing = df_staffing['USER_ID'].nunique()
unique_user_user = df_user['USER_ID'].nunique()
unique_user_xp = df_xp['USER_ID'].nunique()

# Calculate the total number of unique USER_IDs across all tables
total_unique_users = df_user['USER_ID'].nunique()  # Assuming df_user has all unique user ids

# Create a dictionary to store the results
results = {
    "HCK_HEC_LANG": {"Unique Users": unique_user_lang, "Percentage": (unique_user_lang / total_unique_users) * 100},
    "HCK_HEC_SKILLS": {"Unique Users": unique_user_skills, "Percentage": (unique_user_skills / total_unique_users) * 100},
    "HCK_HEC_STAFFING": {"Unique Users": unique_user_staffing, "Percentage": (unique_user_staffing / total_unique_users) * 100},
    "HCK_HEC_USER": {"Unique Users": unique_user_user, "Percentage": (unique_user_user / total_unique_users) * 100},
    "HCK_HEC_XP": {"Unique Users": unique_user_xp, "Percentage": (unique_user_xp / total_unique_users) * 100},
}

# Print the results
for table_name, data in results.items():
    print(f"{table_name}:")
    print(f"  Unique Users: {data['Unique Users']}")
    print(f"  Percentage of Total: {data['Percentage']:.2f}%")


HCK_HEC_LANG:
  Unique Users: 112
  Percentage of Total: 94.12%
HCK_HEC_SKILLS:
  Unique Users: 93
  Percentage of Total: 78.15%
HCK_HEC_STAFFING:
  Unique Users: 119
  Percentage of Total: 100.00%
HCK_HEC_USER:
  Unique Users: 119
  Percentage of Total: 100.00%
HCK_HEC_XP:
  Unique Users: 114
  Percentage of Total: 95.80%


We can see that in some tables the USER_ID is not always unique, suggesting each user can have one or more row entries.


In [5]:
# Set the display options to show all columns
pd.set_option('display.max_rows', None)


In [6]:
df_skills['DOMAIN_DSC'].value_counts()

DOMAIN_DSC
Data Management                    831
Programming Language               724
Data Analytics                     508
Data Integration                   445
Agile Delivery                     365
Data Platform                      208
Financial Close                    116
Financial Planning and Analysis    100
Data Science                        97
Data Governance                     17
Name: count, dtype: int64

In [7]:
df_skills['SKILLS_DSC'].value_counts()

SKILLS_DSC
SQL                                                  188
Snowflake Data Cloud                                 185
Microsoft Power BI                                   170
Microsoft SQL Server                                 136
Python                                               120
Atlassian JIRA Software                              104
DAX                                                   99
Microsoft Azure DevOps Services                       87
Tableau Desktop and Online                            85
Microsoft Azure Data Factory                          68
Microsoft Azure SQL Database                          66
Google BigQuery                                       62
Java                                                  60
Microsoft SQL Server Integration Services (SSIS)      56
R                                                     53
Microstrategy                                         53
PL/SQL                                                52
Microsoft Azure Data Lake Store                       51
DBT                                                   51
T-SQL                                                 50
Microsoft Azure                                       49
PostgreSQL                                            48
Databricks Lakehouse Platform                         47
Project Management                                    46
Javascript                                            46
Oracle Database                                       46
Matillion ETL                                         42
IBM Db2                                               41
Microsoft SQL Server Reporting Services (SSRS)        41
Informatica PowerCenter                               40
GitHub Enterprise                                     39
Microsoft Azure Synapse Analytics                     38
SAP BusinessObjects BI                                37
Fivetran                                              32
Oracle MySQL                                          32
Microsoft Analytics Platform System (SSAS)            32
Microsoft Fabric                                      32
GitLab                                                29
Informatica Intelligent Cloud Services                28
MDX                                                   27
Amazon Web Services                                   26
Airbyte                                               24
Oracle Hyperion Essbase                               23
Google Cloud Platform                                 22
Talend Open Studio                                    22
VBA                                                   22
Dataiku                                               21
Terraform                                             20
Denodo Platform                                       20
Teradata Database                                     20
IBM Netezza Performance Server                        19
CCH Tagetik CPM Planning                              18
IBM Cognos Analytics                                  17
SAP HANA                                              17
Qlik Sense                                            17
SAP Business Planning and Consolidation (SAP BPC)     16
Salesforce Einstein Analytics                         16
MongoDB                                               15
Fluence                                               15
Oracle Cloud EPM                                      14
Oracle Hyperion Planning                              14
Google Data Studio                                    14
GitLab CI                                             14
SnapLogic Intelligent Integration Platform            14
Oracle Business Intelligence Enterprise Edition       13
Looker                                                13
Qlik View                                             12
Spark                                                 12
SAP BW                                                12
Vena Solutions Budgeting, Planning & Forecasting      12

In [8]:
df_xp['MISSION_DSC'].value_counts()

MISSION_DSC
Exhaustive data analysis                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      6
Gérer le budget et produire les rapports d'avancements.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       6
Gestion des risques                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           5
Animation des cérémonies                                                                            

In [9]:
# Assuming you have your DataFrames (df_lang, df_skills, etc.) loaded as before

def null_counts(df, table_name):
  """Calculates null counts for each column in a DataFrame."""
  null_counts = df.isnull().sum()
  print(f"\nNull counts for {table_name}:")
  print(null_counts)

# Get null counts for each table
null_counts(df_lang, "HCK_HEC_LANG")
null_counts(df_skills, "HCK_HEC_SKILLS")
null_counts(df_staffing, "HCK_HEC_STAFFING")
null_counts(df_user, "HCK_HEC_USER")
null_counts(df_xp, "HCK_HEC_XP")



Null counts for HCK_HEC_LANG:
USER_ID               0
LANGUAGE_SKILL_DSC    0
LANGUAGE_SKILL_LVL    8
dtype: int64

Null counts for HCK_HEC_SKILLS:
USER_ID         0
SKILLS_DSC      0
DOMAIN_DSC      0
LEVEL_VAL     256
dtype: int64

Null counts for HCK_HEC_STAFFING:
USER_ID     0
MONTH_1     0
MONTH_2     0
MONTH_3     0
MONTH_4     0
MONTH_5     0
MONTH_6     0
MONTH_7     0
MONTH_8     0
MONTH_9     0
MONTH_10    0
MONTH_11    0
MONTH_12    0
dtype: int64

Null counts for HCK_HEC_USER:
USER_ID       0
LAST_NAME     0
FIRST_NAME    0
ANNEES_XP     0
dtype: int64

Null counts for HCK_HEC_XP:
USER_ID        0
MISSION_DSC    0
dtype: int64


In [10]:
# Find non-unique rows in df_staffing based on USER_ID
duplicate_user_ids = df_staffing[df_staffing.duplicated(subset=['USER_ID'], keep=False)]['USER_ID'].unique()

# Filter the DataFrame to include only rows with duplicate USER_IDs
df_duplicates = df_staffing[df_staffing['USER_ID'].isin(duplicate_user_ids)]

# Check if df_duplicates is empty before proceeding
if df_duplicates.empty:
    print("No duplicate USER_IDs found in df_staffing.")
    df_inconsistent_months = []  # Initialize as an empty list to avoid errors later
else:
    # Group by USER_ID and check if all values for month columns are the same
    def has_varying_months(group):
        month_cols = [f'MONTH_{i}' for i in range(1, 13)]
        return not group[month_cols].nunique().eq(1).all()

    # Apply the function to each group and filter for cases where months are not the same
    df_inconsistent_months = df_duplicates.groupby('USER_ID').apply(has_varying_months).loc[lambda x: x].index.tolist()

# Filter the DataFrame to show only rows with inconsistent months for the duplicate USER_IDs
df_inconsistent_staffing = df_staffing[df_staffing['USER_ID'].isin(df_inconsistent_months)]

print("Number of rows with non unique USER_IDs:", df_duplicates.shape[0])
print("Number of rows with non unique USER_IDs and inconsistent month columns:", df_inconsistent_staffing.shape[0])


No duplicate USER_IDs found in df_staffing.
Number of rows with non unique USER_IDs: 0
Number of rows with non unique USER_IDs and inconsistent month columns: 0


-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

----------------------------------------------------------------------- **Merging By Ashraf For Skills & Language** ---------------------------------------------------------------

Check for duplicates to know how to deal with them when pivoting the table

In [11]:
# Check for duplicate skill entries per user
duplicate_skills = df_skills.groupby(["USER_ID", "SKILLS_DSC"]).size()
duplicate_skills = duplicate_skills[duplicate_skills > 1]

if not duplicate_skills.empty:
    print("Found duplicate skill entries:")
    print(duplicate_skills)
else:
    print("No duplicate skill entries found.")


Found duplicate skill entries:
USER_ID  SKILLS_DSC                                       
2433083  DBT                                                   4
         Microsoft Analytics Platform System (SSAS)            4
         Microsoft Azure Data Factory                          4
         Microsoft Azure Data Lake Store                       4
         Microsoft Azure SQL Database                          4
         Microsoft SQL Server                                  4
         Microsoft SQL Server Integration Services (SSIS)      4
         PostgreSQL                                            4
         Snowflake Data Cloud                                  4
2433092  Atlassian JIRA Software                               6
         Microsoft Azure DevOps Services                       6
2433099  Atlassian JIRA Software                              10
         Microsoft Azure DevOps Services                      10
         SAP BW                                               10


In [14]:
# Find duplicated (USER_ID, SKILLS_DSC) combinations
dupe_skill_keys = df_skills.duplicated(subset=["USER_ID", "SKILLS_DSC"], keep=False)

# Show only the duplicated rows
duplicate_skill_entries = df_skills[dupe_skill_keys]

print(duplicate_skill_entries.sort_values(by=["USER_ID", "SKILLS_DSC"]))


      USER_ID                                         SKILLS_DSC  \
2552  2433083                                                DBT   
2555  2433083                                                DBT   
2558  2433083                                                DBT   
2599  2433083                                                DBT   
2254  2433083         Microsoft Analytics Platform System (SSAS)   
2260  2433083         Microsoft Analytics Platform System (SSAS)   
2267  2433083         Microsoft Analytics Platform System (SSAS)   
2273  2433083         Microsoft Analytics Platform System (SSAS)   
2553  2433083                       Microsoft Azure Data Factory   
2556  2433083                       Microsoft Azure Data Factory   
2597  2433083                       Microsoft Azure Data Factory   
2600  2433083                       Microsoft Azure Data Factory   
2253  2433083                    Microsoft Azure Data Lake Store   
2259  2433083                    Microsoft Azure

In [12]:
# Check for duplicate language entries per user
duplicate_lang = df_lang.groupby(["USER_ID", "LANGUAGE_SKILL_DSC"]).size()
duplicate_lang = duplicate_lang[duplicate_lang > 1]

if not duplicate_lang.empty:
    print("Found duplicate language entries:")
    print(duplicate_lang)
else:
    print("No duplicate language entries found.")


Found duplicate language entries:
USER_ID  LANGUAGE_SKILL_DSC
2503861  English               2
dtype: int64


In [13]:
# Find duplicated (USER_ID, LANGUAGE_SKILL_DSC) combinations
dupe_lang_keys = df_lang.duplicated(subset=["USER_ID", "LANGUAGE_SKILL_DSC"], keep=False)

# Show only the duplicated rows
duplicate_lang_entries = df_lang[dupe_lang_keys]

print(duplicate_lang_entries.sort_values(by=["USER_ID", "LANGUAGE_SKILL_DSC"]))


    USER_ID LANGUAGE_SKILL_DSC  LANGUAGE_SKILL_LVL
8   2503861            English                 0.0
78  2503861            English                90.0


All duplicates are either 0, Null or the same value. We will pivot the tables on the highest value.

Joining tables below

In [24]:
import pandas as pd

def pivot_skills_table(df_skills: pd.DataFrame) -> pd.DataFrame:
    """
    Pivot the skills table to ensure all skills are included, even if no current user has them.
    """
    # Step 1: Get all unique skills
    all_skills = df_skills["SKILLS_DSC"].unique()

    # Step 2: Create dummy rows with a fake user to ensure all skills appear
    dummy_user_id = -99999
    dummy_rows = pd.DataFrame({
        "USER_ID": [dummy_user_id] * len(all_skills),
        "SKILLS_DSC": all_skills,
        "LEVEL_VAL": [0] * len(all_skills)
    })

    # Step 3: Append dummy to original skills table
    df_skills_augmented = pd.concat([df_skills, dummy_rows], ignore_index=True)

    # Step 4: Pivot
    skills_pivot = df_skills_augmented.pivot_table(
        index="USER_ID",
        columns="SKILLS_DSC",
        values="LEVEL_VAL",
        aggfunc='max',
        fill_value=0
    )

    # Step 5: Drop the dummy user and rename columns
    skills_pivot = skills_pivot.drop(index=dummy_user_id)
    skills_pivot.columns = [f"{col}_level" for col in skills_pivot.columns]

    return skills_pivot.reset_index()

def pivot_language_table(df_lang: pd.DataFrame) -> pd.DataFrame:
    """
    Pivot the language table to ensure English and French columns are always included,
    even if no users currently have those language scores.
    """
    # Define the languages you want to ensure are always present
    required_languages = ["English", "French"]

    # Step 1: Create dummy rows for each language
    dummy_user_id = -99999
    dummy_rows = pd.DataFrame({
        "USER_ID": [dummy_user_id] * len(required_languages),
        "LANGUAGE_SKILL_DSC": required_languages,
        "LANGUAGE_SKILL_LVL": [0] * len(required_languages)
    })

    # Step 2: Append dummy to original lang table
    df_lang_augmented = pd.concat([df_lang, dummy_rows], ignore_index=True)

    # Step 3: Pivot
    lang_pivot = df_lang_augmented.pivot_table(
        index="USER_ID",
        columns="LANGUAGE_SKILL_DSC",
        values="LANGUAGE_SKILL_LVL",
        aggfunc='max',
        fill_value=0
    )

    # Step 4: Drop the dummy user and rename columns
    lang_pivot = lang_pivot.drop(index=dummy_user_id, errors='ignore')
    lang_pivot = lang_pivot.rename(columns={
        "English": "English_level",
        "French": "French_level"
    })

    return lang_pivot.reset_index()


def merge_user_skills_languages(
    df_user: pd.DataFrame,
    skills_pivot: pd.DataFrame,
    lang_pivot: pd.DataFrame
) -> pd.DataFrame:
    """
    Merges the user base table with pivoted skills and language tables.
    Fills missing values with 0.
    """
    merged_df = df_user.merge(skills_pivot, on="USER_ID", how="left")
    merged_df = merged_df.merge(lang_pivot, on="USER_ID", how="left")
    merged_df.fillna(0, inplace=True)
    return merged_df


skills_pivoted = pivot_skills_table(df_skills)
lang_pivoted = pivot_language_table(df_lang)
Users_Lang_Skills = merge_user_skills_languages(df_user, skills_pivoted, lang_pivoted)

# Optional: Save
# final_df.to_csv("merged_user_skills_languages.csv", index=False)
Users_Lang_Skills.head(10)

USER_ID LAST_NAME FIRST_NAME  ANNEES_XP  Airbyte_level  \
0  2843838     Ramos      David        1.0            0.0   
1  2479537     Small       Carl        2.5            0.0   
2  2533337     Evans     Carmen        2.0            0.0   
3  2446382   Pittman     Brandi        2.5            0.0   
4  2433124    Thomas      Julie        2.5            0.0   
5  2817191   Johnson       Carl        1.0            0.0   
6  2936770      Choi    Shannon        0.5            0.0   
7  2501258  Martinez      Brett        2.5            0.0   
8  2433139    Holden   Jeremiah        2.5            0.0   
9  2480827   Hubbard      Megan        2.5            0.0   

   Alteryx Analytics Hub_level  Alteryx Designer_level  \
0                          0.0                     0.0   
1                          0.0                     0.0   
2                          0.0                     0.0   
3                          0.0                     0.0   
4                          0.0                     0.0   
5                          0.0                     0.0   
6                          0.0                     0.0   
7                          0.0                     0.0   
8                          0.0                     0.0   
9                          0.0                     0.0   

   Amazon AWS CloudFormation_level  Amazon AWS Glue_level  \
0                              0.0                    0.0   
1                              0.0                    0.0   
2                              0.0                    0.0   
3                              0.0                    0.0   
4                              0.0                    0.0   
5                              0.0                    0.0   
6                              0.0                    0.0   
7                              0.0                    0.0   
8                              0.0                    0.0   
9                              0.0                    0.0   

   Amazon QuickSight_level  ...  Teradata Vantage_level  Terraform_level  \
0                      0.0  ...                     0.0              0.0   
1                      0.0  ...                     0.0              0.0   
2                      0.0  ...                     0.0              0.0   
3                      0.0  ...                     0.0              0.0   
4                      0.0  ...                     0.0              0.0   
5                      0.0  ...                     0.0              0.0   
6                      0.0  ...                     0.0              0.0   
7                      0.0  ...                     0.0              0.0   
8                      0.0  ...                     0.0              0.0   
9                      0.0  ...                     0.0              0.0   

   VBA_level  Vena Close_level  \
0        0.0               0.0   
1      100.0               0.0   
2        0.0               0.0   
3        0.0               0.0   
4        0.0               0.0   
5        0.0               0.0   
6        0.0               0.0   
7        0.0               0.0   
8        0.0               0.0   
9        0.0               0.0   

   Vena Solutions Budgeting, Planning & Forecasting_level  \
0                                                0.0        
1                                                0.0        
2                                                0.0        
3                                                0.0        
4                                                0.0        
5                                                0.0        
6                                                0.0        
7                                                0.0        
8                                                0.0        
9                                                0.0        

   Workday Adaptive Planning_level  Workiva_level  \
0                              0.0            0.0   
1                             

In [25]:
Users_Lang_Skills.shape

(119, 139)

In [26]:
num_unique_skills = df_skills["SKILLS_DSC"].nunique()
print(f"Number of unique skills: {num_unique_skills}")

Number of unique skills: 133


The code below checks if we missed any skills

In [27]:
# 1. Get all unique skill names from the skills table
all_skills = df_skills["SKILLS_DSC"].unique()
all_skill_columns = set(f"{skill}_level" for skill in all_skills)

# 2. Get the actual skill columns in the merged dataset
user_skill_columns = set(col for col in Users_Lang_Skills.columns if col.endswith("_level")
                         and not col.startswith(("English", "French")))

# 3. Find missing skills
missing_skills = all_skill_columns - user_skill_columns

print("Missing skills in merged table:")
print(missing_skills)



Missing skills in merged table:
set()


---------------------------------------------------------- **Mapping Skills/Languages/Experience efforts** ----------------------------

In [35]:
# List of skills to check
skills_to_check = [
    "Spark"
]

# Get unique skills from the dataset
existing_skills = df_skills["SKILLS_DSC"].unique()

# Check which are present and which are missing
present_skills = [skill for skill in skills_to_check if skill in existing_skills]
missing_skills = [skill for skill in skills_to_check if skill not in existing_skills]

# Output
print("✅ Present skills:")
print(present_skills)

print("\n❌ Missing skills:")
print(missing_skills)


✅ Present skills:
['Spark']

❌ Missing skills:
[]


In [31]:
# Get unique skill names, remove any NaNs, and sort them alphabetically
unique_skills_sorted = sorted(df_skills["SKILLS_DSC"].dropna().unique())

# Display the list
for skill in unique_skills_sorted:
    print(skill)


Airbyte
Alteryx Analytics Hub
Alteryx Designer
Amazon AWS CloudFormation
Amazon AWS Glue
Amazon QuickSight
Amazon Redshift
Amazon Web Services
Anaconda Enterprise
Anaplan for Finance
Applied OLAP Dodeca
Atlassian Bitbucket
Atlassian JIRA Software
Azure Purview
CCH Tagetik CPM Conso
CCH Tagetik CPM Planning
Cloudera Data Platform
Cloudera Data Warehouse
Cloudera Enterprise Data Hub
Colibra
DAX
DBT
Data Galaxy
Databricks
Databricks Lakehouse Platform
Dataiku
Denodo Platform
Docker
Domo
Fivetran
Fluence
GitHub Enterprise
GitLab
GitLab CI
Google BigQuery
Google Cloud DataProc
Google Cloud Platform
Google Data Studio
Google Vertex AI
Hadoop
IBM Cognos Analytics
IBM Cognos TM1
IBM DataStage
IBM Db2
IBM Netezza Performance Server
IBM SPSS Statistics
Informatica Data Quality (IDQ)
Informatica Intelligent Cloud Services
Informatica PowerCenter
Java
Javascript
KNIME Analytics Platform
Kubernetes
Looker
MDX
Matillion ETL
Matlab
Microsoft Analytics Platform System (SSAS)
Microsoft Azure
Microsoft 

In [37]:
# !pip install fuzzywuzzy

**Matching extracted skill with matched skill**

Used FuzzyWuzzy with 90% threshold i found gave best results for data engineering.

In [58]:
import pandas as pd
import ast
from fuzzywuzzy import process

# Load the extracted features CSV
df_features = pd.read_csv("extracted_features.csv")

# === Specify the job you want to analyze ===
selected_job = "Data Engineer"  # ← Change this to your demand_id of interest

# Filter the row for the selected job
selected_row = df_features[df_features["demand_id"] == selected_job]

if selected_row.empty:
    raise ValueError(f"No job found with demand_id: {selected_job}")

# Parse the 'extracted_features' column
parsed = ast.literal_eval(selected_row.iloc[0]["extracted_features"])

# Extract the required skills list
required_skills = parsed.get("Required Skills", [])

# Standardize skill names
def standardize_skill_name(name):
    return name.lower().strip().replace("-", " ").replace("_", " ")

extracted_skills_std = [standardize_skill_name(skill) for skill in required_skills]

# === Replace this with actual known skill names from your pivot table ===
known_skills = [col.replace("_level", "") for col in skills_pivoted.columns]

# Fuzzy matching
def match_extracted_skill(extracted_skill, known_skills, threshold=90):
    match, score = process.extractOne(extracted_skill, known_skills)
    return match if score >= threshold else None

# Create the mapping
mapping = {}
for skill in extracted_skills_std:
    match = match_extracted_skill(skill, known_skills)
    if match:
        mapping[skill] = match

# Output as DataFrame
mapping_df = pd.DataFrame(list(mapping.items()), columns=["Extracted Skill", "Matched Skill"])
print(mapping_df)

      Extracted Skill                 Matched Skill
0        apache spark                         Spark
1              python                        Python
2               scala                         Scala
3  azure data factory  Microsoft Azure Data Factory
4    azure databricks                    Databricks


Added matched_skill_level, which is just the extracted skills name from the job description and the level is the matched skills value.

In [60]:
def add_extracted_skill_scores_to_users() -> pd.DataFrame:
    """
    Adds <extracted_skill>_score columns to Users_Lang_Skills based on the mapping dict.
    Each score column is populated from the matched skill's _level column.
    NaNs are filled with 0 for users who do not have the skill.

    Returns:
        A new DataFrame with *_score columns added.
    """
    df = Users_Lang_Skills.copy()

    for extracted_skill, matched_skill in mapping.items():
        matched_col = f"{matched_skill}_level"
        new_score_col = f"{extracted_skill}_score"

        if matched_col in df.columns:
            df[new_score_col] = df[matched_col].fillna(0)
        else:
            df[new_score_col] = 0  # If matched skill column doesn't exist

    return df


Next code will add the scores into a new dataframe

In [61]:
Users_Lang_Skills_Scored = add_extracted_skill_scores_to_users()

In [62]:
# First, get all the score columns created from extracted skills
score_columns = [f"{skill}_score" for skill in mapping.keys()]

# Filter users who have at least one non-zero score in any of those columns
users_with_matched_skills = Users_Lang_Skills_Scored[
    Users_Lang_Skills_Scored[score_columns].sum(axis=1) > 0
]

# Preview the results
print(users_with_matched_skills[["USER_ID", "FIRST_NAME", "LAST_NAME"] + score_columns])


     USER_ID FIRST_NAME  LAST_NAME  apache spark_score  python_score  \
0    2843838      David      Ramos                 0.0          80.0   
2    2533337     Carmen      Evans                 0.0          40.0   
3    2446382     Brandi    Pittman                 0.0          60.0   
16   2433117    Clayton      Black                 0.0          20.0   
19   2445183  Alexandra   Valencia                 0.0           0.0   
20   2671704      Nancy  Blackburn                 0.0          90.0   
24   2452725   Jennifer     Harris                 0.0         100.0   
26   2583132     Shelby      Mcgee                 0.0          70.0   
27   2513866     Samuel      Jones                 0.0          80.0   
28   2443989       Tina    Hopkins                 0.0          40.0   
29   2700124    Melissa      Kelly                 0.0          50.0   
30   2836571    Jeffrey     Foster                 0.0          60.0   
35   2443199     Sherry     Howell                 0.0          

**Adding Language Scores**

In [72]:
def add_extracted_language_scores_to_users(selected_job: str) -> pd.DataFrame:
    """
    Adds <language>_lang_score columns to Users_Lang_Skills_Scored for a given job,
    based on required languages extracted from extracted_features.csv.

    It uses fuzzy matching to align extracted languages with actual user columns
    (e.g., French_level → french_lang_score).

    Parameters:
        selected_job (str): The job title (demand_id) to extract languages from.

    Returns:
        pd.DataFrame: Updated Users_Lang_Skills_Scored with *_lang_score columns added.
    """
    import ast
    from fuzzywuzzy import process

    # Step 1: Select row for the specified job
    selected_row = df_features[df_features["demand_id"] == selected_job]
    if selected_row.empty:
        raise ValueError(f"No job found with demand_id: {selected_job}")

    parsed = ast.literal_eval(selected_row.iloc[0]["extracted_features"])
    required_languages = parsed.get("Languages", [])

    # Step 2: Standardize language names
    def standardize_lang_name(name):
        return name.lower().strip()

    required_languages_std = [standardize_lang_name(lang) for lang in required_languages]

    # Step 3: Get actual column names of known language levels
    known_langs = {
        col.replace("_level", "").lower(): col
        for col in Users_Lang_Skills_Scored.columns
        if col.endswith("_level") and col.replace("_level", "").lower() in ["english", "french"]
    }

    # Step 4: Match extracted languages to known columns
    lang_mapping = {}
    for lang in required_languages_std:
        match, score = process.extractOne(lang, list(known_langs.keys()))
        if score >= 85:
            lang_mapping[lang] = known_langs[match]

    # Step 5: Add *_lang_score columns
    df = Users_Lang_Skills_Scored.copy()
    for extracted_lang, matched_col in lang_mapping.items():
        new_score_col = f"{extracted_lang}_lang_score"
        df[new_score_col] = df[matched_col].fillna(0) if matched_col in df.columns else 0

    return df



In [73]:
Users_Lang_Skills_Scored = add_extracted_language_scores_to_users("Data Engineer")

In [74]:
# Show summary stats for new language score columns
lang_score_columns = [f"{lang}_lang_score" for lang in lang_mapping.keys()]

print("Summary of language match scores:")
print(Users_Lang_Skills_Scored[lang_score_columns].describe())


Summary of language match scores:
       french_lang_score  english_lang_score
count         119.000000          119.000000
mean           81.764706           75.798319
std            38.237595           32.061799
min             0.000000            0.000000
25%           100.000000           65.000000
50%           100.000000           80.000000
75%           100.000000          100.000000
max           100.000000          100.000000


In [75]:
# Filter users with at least one non-zero language match score
users_with_matched_languages = Users_Lang_Skills_Scored[
    Users_Lang_Skills_Scored[lang_score_columns].sum(axis=1) > 0
]

# Display selected user info + their matched language scores
print(users_with_matched_languages[["USER_ID", "FIRST_NAME", "LAST_NAME"] + lang_score_columns])


     USER_ID   FIRST_NAME  LAST_NAME  french_lang_score  english_lang_score
0    2843838        David      Ramos              100.0               100.0
1    2479537         Carl      Small              100.0                90.0
2    2533337       Carmen      Evans              100.0                50.0
3    2446382       Brandi    Pittman              100.0                60.0
4    2433124        Julie     Thomas              100.0               100.0
5    2817191         Carl    Johnson              100.0               100.0
7    2501258        Brett   Martinez              100.0               100.0
8    2433139     Jeremiah     Holden              100.0               100.0
9    2480827        Megan    Hubbard                0.0                90.0
10   2433122         Mary        Cox              100.0                80.0
11   2433125       Marvin      Smith              100.0               100.0
12   2513848       Dustin      Brown              100.0                80.0
13   2433130

In [76]:
Users_Lang_Skills_Scored.head()

USER_ID LAST_NAME FIRST_NAME  ANNEES_XP  Airbyte_level  \
0  2843838     Ramos      David        1.0            0.0   
1  2479537     Small       Carl        2.5            0.0   
2  2533337     Evans     Carmen        2.0            0.0   
3  2446382   Pittman     Brandi        2.5            0.0   
4  2433124    Thomas      Julie        2.5            0.0   

   Alteryx Analytics Hub_level  Alteryx Designer_level  \
0                          0.0                     0.0   
1                          0.0                     0.0   
2                          0.0                     0.0   
3                          0.0                     0.0   
4                          0.0                     0.0   

   Amazon AWS CloudFormation_level  Amazon AWS Glue_level  \
0                              0.0                    0.0   
1                              0.0                    0.0   
2                              0.0                    0.0   
3                              0.0                    0.0   
4                              0.0                    0.0   

   Amazon QuickSight_level  ...  insightsoftware Longview Planning_level  \
0                      0.0  ...                                      0.0   
1                      0.0  ...                                      0.0   
2                      0.0  ...                                      0.0   
3                      0.0  ...                                      0.0   
4                      0.0  ...                                      0.0   

   English_level  French_level  apache spark_score  python_score  scala_score  \
0          100.0         100.0                 0.0          80.0          0.0   
1           90.0         100.0                 0.0           0.0          0.0   
2           50.0         100.0                 0.0          40.0          0.0   
3           60.0         100.0                 0.0          60.0          0.0   
4          100.0         100.0                 0.0           0.0          0.0   

   azure data factory_score  azure databricks_score  french_lang_score  \
0                       0.0                     0.0              100.0   
1                       0.0                     0.0              100.0   
2                       0.0                     0.0              100.0   
3                       0.0                     0.0              100.0   
4                       0.0                     0.0              100.0   

   english_lang_score  
0               100.0  
1                90.0  
2                50.0  
3                60.0  
4               100.0  

[5 rows x 146 columns]

**Adding Experience scores**

100 for experience equal to or greater than required, 0 otherwise

Found that we dont have any candidates with 5 years of experience. What should our next steps be?

In [77]:
def add_extracted_experience_score_to_users(selected_job: str) -> pd.DataFrame:
    """
    Adds an 'experience_score' column to Users_Lang_Skills_Scored based on the
    'Experience Required' from the extracted_features for the selected job.

    - If ANNEES_XP >= Experience Required → score = 100
    - Else → score = 0

    Parameters:
        selected_job (str): The job title (demand_id) to extract the required experience for.

    Returns:
        pd.DataFrame: Updated Users_Lang_Skills_Scored with 'experience_score' column added.
    """
    import ast

    # Step 1: Select job row
    selected_row = df_features[df_features["demand_id"] == selected_job]
    if selected_row.empty:
        raise ValueError(f"No job found with demand_id: {selected_job}")

    parsed = ast.literal_eval(selected_row.iloc[0]["extracted_features"])
    required_experience = parsed.get("Experience Required", 0)

    # Step 2: Calculate experience_score
    df = Users_Lang_Skills_Scored.copy()
    df["experience_score"] = df["ANNEES_XP"].apply(
        lambda x: 100 if x >= required_experience else 0
    )

    return df


In [78]:
Users_Lang_Skills_Scored = add_extracted_experience_score_to_users("Data Engineer")

In [79]:
print("Experience Score Summary:")
print(Users_Lang_Skills_Scored["experience_score"].describe())


Experience Score Summary:
count    119.0
mean       0.0
std        0.0
min        0.0
25%        0.0
50%        0.0
75%        0.0
max        0.0
Name: experience_score, dtype: float64


In [80]:
import ast

selected_job = "Data Engineer"  # or whichever job you're testing
selected_row = df_features[df_features["demand_id"] == selected_job]
parsed = ast.literal_eval(selected_row.iloc[0]["extracted_features"])
required_experience = parsed.get("Experience Required", 0)

print(f"Experience Required for '{selected_job}': {required_experience} years")


Experience Required for 'Data Engineer': 5.0 years


In [81]:
print("User experience (ANNEES_XP) summary:")
print(Users_Lang_Skills_Scored["ANNEES_XP"].describe())


User experience (ANNEES_XP) summary:
count    119.000000
mean       2.029412
std        0.655143
min        0.500000
25%        1.500000
50%        2.500000
75%        2.500000
max        2.500000
Name: ANNEES_XP, dtype: float64


In [83]:
print("Original user experience summary from df_user:")
print(df_user["ANNEES_XP"].describe())

# Optionally, see the most experienced users
print("\nUsers with highest experience:")
print(df_user.sort_values("ANNEES_XP", ascending=False).head(10)[["USER_ID", "FIRST_NAME", "LAST_NAME", "ANNEES_XP"]])

Original user experience summary from df_user:
count    119.000000
mean       2.029412
std        0.655143
min        0.500000
25%        1.500000
50%        2.500000
75%        2.500000
max        2.500000
Name: ANNEES_XP, dtype: float64

Users with highest experience:
    USER_ID FIRST_NAME LAST_NAME  ANNEES_XP
1   2479537       Carl     Small        2.5
4   2433124      Julie    Thomas        2.5
3   2446382     Brandi   Pittman        2.5
10  2433122       Mary       Cox        2.5
9   2480827      Megan   Hubbard        2.5
8   2433139   Jeremiah    Holden        2.5
7   2501258      Brett  Martinez        2.5
19  2445183  Alexandra  Valencia        2.5
16  2433117    Clayton     Black        2.5
17  2433135     Dakota   Nichols        2.5


**Preferred skills**

In [84]:
def add_extracted_preferred_skill_scores_to_users(selected_job: str) -> pd.DataFrame:
    """
    Adds <preferred_skill>_preferred_score columns to Users_Lang_Skills_Scored based on
    the 'Preferred Skills' extracted from a job description. Uses fuzzy matching to match
    extracted skills to known skill columns (e.g., 'python_level').

    Parameters:
        selected_job (str): Job title (demand_id) to extract preferred skills from.

    Returns:
        pd.DataFrame: Updated Users_Lang_Skills_Scored with *_preferred_score columns added.
    """
    import ast
    from fuzzywuzzy import process

    # Step 1: Get the job description
    selected_row = df_features[df_features["demand_id"] == selected_job]
    if selected_row.empty:
        raise ValueError(f"No job found with demand_id: {selected_job}")

    parsed = ast.literal_eval(selected_row.iloc[0]["extracted_features"])
    preferred_skills = parsed.get("Preferred Skills", [])

    # Step 2: Standardize extracted skill names
    def standardize_skill_name(name):
        return name.lower().strip().replace("-", " ").replace("_", " ")

    preferred_skills_std = [standardize_skill_name(skill) for skill in preferred_skills]

    # Step 3: Get known skill names from Users_Lang_Skills_Scored columns
    known_skills = [col.replace("_level", "") for col in Users_Lang_Skills_Scored.columns if col.endswith("_level")]

    # Step 4: Fuzzy match extracted preferred skills to known skills
    mapping = {}
    for skill in preferred_skills_std:
        match, score = process.extractOne(skill, known_skills)
        if score >= 85:
            mapping[skill] = match

    # Step 5: Create *_preferred_score columns
    df = Users_Lang_Skills_Scored.copy()
    for extracted_skill, matched_skill in mapping.items():
        matched_col = f"{matched_skill}_level"
        new_score_col = f"{extracted_skill}_preferred_score"

        if matched_col in df.columns:
            df[new_score_col] = df[matched_col].fillna(0)
        else:
            df[new_score_col] = 0

    return df


In [85]:
Users_Lang_Skills_Scored = add_extracted_preferred_skill_scores_to_users("Data Engineer")

Sanity check of transformation

In [86]:
import ast
from fuzzywuzzy import process

# === Step 1: Select the job
selected_job = "Data Engineer"  # Change as needed
selected_row = df_features[df_features["demand_id"] == selected_job]

parsed = ast.literal_eval(selected_row.iloc[0]["extracted_features"])
preferred_skills = parsed.get("Preferred Skills", [])

# === Step 2: Standardize extracted skills
def standardize_skill_name(name):
    return name.lower().strip().replace("-", " ").replace("_", " ")

preferred_skills_std = [standardize_skill_name(skill) for skill in preferred_skills]

# === Step 3: Get known skill names from the user dataframe
known_skills = [col.replace("_level", "") for col in Users_Lang_Skills_Scored.columns if col.endswith("_level")]

# === Step 4: Build mapping with score
mapped_results = []
for skill in preferred_skills_std:
    match, score = process.extractOne(skill, known_skills)
    mapped_results.append((skill, match, score))

# === Step 5: Display mapping
mapping_df = pd.DataFrame(mapped_results, columns=["Extracted Preferred Skill", "Matched Skill", "Score"])
print(mapping_df)


Empty DataFrame
Columns: [Extracted Preferred Skill, Matched Skill, Score]
Index: []


In [87]:
import ast

selected_job = "Data Engineer"  # or whichever you're using
row = df_features[df_features["demand_id"] == selected_job]

if row.empty:
    print(f"No job found with demand_id: {selected_job}")
else:
    parsed = ast.literal_eval(row.iloc[0]["extracted_features"])
    preferred_skills = parsed.get("Preferred Skills", [])
    print(f"Preferred Skills for '{selected_job}':", preferred_skills)


Preferred Skills for 'Data Engineer': []


Data Engineer has no preferred skills

Testing on data analyst for preferred skills

It works fine. threshold between 85 to 90 are usually best.

Trouble is found with Scrum job description. The required skills format for that job are funny

In [100]:
# import ast
# from fuzzywuzzy import process
# import pandas as pd

# # === Step 1: Choose the job to test
# selected_job = "Data Analyst"  # <-- Change this to test another job
# threshold = 85        # <-- Adjust threshold here

# # Step 2: Extract preferred skills from that job
# row = df_features[df_features["demand_id"] == selected_job]

# if row.empty:
#     print(f"No job found with demand_id: {selected_job}")
# else:
#     parsed = ast.literal_eval(row.iloc[0]["extracted_features"])
#     preferred_skills = parsed.get("Preferred Skills", [])

#     if not preferred_skills:
#         print(f"No preferred skills found for '{selected_job}'")
#     else:
#         # Step 3: Standardize preferred skill names
#         def standardize_skill_name(name):
#             return name.lower().strip().replace("-", " ").replace("_", " ")

#         preferred_skills_std = [standardize_skill_name(skill) for skill in preferred_skills]

#         # Step 4: Get known skills from current user skill columns
#         known_skills = [col.replace("_level", "") for col in Users_Lang_Skills_Scored.columns if col.endswith("_level")]

#         # Step 5: Fuzzy match each preferred skill with threshold filtering
#         mapped_results = []
#         for skill in preferred_skills_std:
#             match, score = process.extractOne(skill, known_skills)
#             if score >= threshold:
#                 mapped_results.append((skill, match, score))

#         # Step 6: Show filtered mapping
#         if mapped_results:
#             mapping_df = pd.DataFrame(mapped_results, columns=["Extracted Preferred Skill", "Matched Skill", "Score"])
#             print(mapping_df)
#         else:
#             print(f"No preferred skills matched with threshold ≥ {threshold}")



  Extracted Preferred Skill       Matched Skill  Score
0                       dax                 DAX    100
1               power query  Microsoft Power BI     86
2                       etl       Matillion ETL     90


**Adding the availability score**

which is derived from the staffing table with months. If a job specifies 10 months, the candidate should have 0 in at least 10 columns in staffing. Since each column is month availability.

This function will be applied to Data Analyst since its the only job with duration

In [120]:
def add_availability_score_to_users(selected_job: str) -> pd.DataFrame:
    """
    Adds an 'availability_score' column to Users_Lang_Skills_Scored based on the
    'Duration' from the job description and availability in HCK_HEC_STAFFING.

    A user gets a score of 100 if they are fully available (value == 0)
    for at least 'Duration' number of months, otherwise 0.

    Parameters:
        selected_job (str): The demand_id of the job to evaluate.

    Returns:
        pd.DataFrame: Updated Users_Lang_Skills_Scored with 'availability_score'.
    """
    import ast
    import re

    # Step 1: Get the job row based on selected_job
    job_row = df_features[df_features["demand_id"] == selected_job]
    if job_row.empty:
        raise ValueError(f"No job found with demand_id: {selected_job}")

    # Step 2: Extract the "Additional Notes" field to get the "Duration"
    extracted_features = ast.literal_eval(job_row.iloc[0]["extracted_features"])
    additional_notes = extracted_features.get("Additional Notes", "")

    # Step 3: Use regex to extract the duration from "Additional Notes"
    duration_match = re.search(r"Duration:\s*(\d+)\s*months", additional_notes)
    if duration_match:
        duration = int(duration_match.group(1))  # Extract and convert to integer
    else:
        duration = 0  # Default to 0 if not found

    print(f"Required duration for '{selected_job}': {duration} months")  # Debugging step

    # Step 4: Define month columns (MONTH_1 to MONTH_12)
    month_cols = [col for col in df_staffing.columns if col.startswith("MONTH_")]

    # Step 5: Count available months (value == 0) for each user
    staffing_availability = df_staffing.copy()
    staffing_availability["available_months"] = (staffing_availability[month_cols] == 0).sum(axis=1)

    # Step 6: Check the available months distribution for debugging
    print(staffing_availability["available_months"].describe())  # Debugging step

    # Step 7: Determine availability score based on the extracted duration
    staffing_availability["availability_score"] = staffing_availability["available_months"].apply(
        lambda x: 100 if x >= duration else 0
    )

    # Step 8: Merge the availability_score into Users_Lang_Skills_Scored without modifying the original DataFrame
    df = Users_Lang_Skills_Scored.copy()
    df = df.merge(
        staffing_availability[["USER_ID", "availability_score"]],
        on="USER_ID",
        how="left"
    )
    df["availability_score"].fillna(0, inplace=True)  # Handle missing availability scores

    return df

In [121]:
Users_availability = add_availability_score_to_users("Data Analyst")
print(Users_availability[["USER_ID", "availability_score"]].head())

Required duration for 'Data Analyst': 10 months
count    119.000000
mean       2.579832
std        4.770133
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max       12.000000
Name: available_months, dtype: float64
   USER_ID  availability_score
0  2843838                   0
1  2479537                   0
2  2533337                   0
3  2446382                   0
4  2433124                   0


<ipython-input-120-b8bd8c39bff4>:58: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["availability_score"].fillna(0, inplace=True)  # Handle missing availability scores


In [122]:
# Count number of users with availability_score of 100
num_users_with_100_score = Users_availability[Users_availability["availability_score"] == 100].shape[0]

# Count number of users with availability_score of 0
num_users_with_0_score = Users_availability[Users_availability["availability_score"] == 0].shape[0]

# Print the results
print(f"Number of users with availability score of 100: {num_users_with_100_score}")
print(f"Number of users with availability score of 0: {num_users_with_0_score}")


Number of users with availability score of 100: 26
Number of users with availability score of 0: 93


Check if the availability result matches the original staffing dataframe

In [124]:
# Step 1: Calculate available months (value == 0) for each user in the original dataset
staffing_availability = df_staffing.copy()

# Define month columns (representing months 1 to 12)
month_cols = [f"MONTH_{i}" for i in range(1, 13)]
staffing_availability["available_months"] = (staffing_availability[month_cols] == 0).sum(axis=1)

# Step 2: Users with at least 10 available months (0% unavailable)
users_with_at_least_10_available_months_orig = staffing_availability[staffing_availability["available_months"] >= 10]
num_users_with_100_score_orig = users_with_at_least_10_available_months_orig.shape[0]

# Step 3: Users with fewer than 10 available months (availability score = 0)
users_without_10_available_months_orig = staffing_availability[staffing_availability["available_months"] < 10]
num_users_with_0_score_orig = users_without_10_available_months_orig.shape[0]

# Step 4: Compare with Users_availability results
print(f"Number of users with availability score of 100 (from Users_availability): {num_users_with_100_score}")
print(f"Number of users with availability score of 0 (from Users_availability): {num_users_with_0_score}")

# Step 5: Compare with original dataset
print(f"Number of users with at least 10 available months (original dataset): {num_users_with_100_score_orig}")
print(f"Number of users with fewer than 10 available months (original dataset): {num_users_with_0_score_orig}")

# Step 6: Verify if the numbers match
if num_users_with_100_score == num_users_with_100_score_orig and num_users_with_0_score == num_users_with_0_score_orig:
    print("The counts match between Users_availability and the original dataset.")
else:
    print("The counts do not match. Please check the calculations.")


Number of users with availability score of 100 (from Users_availability): 26
Number of users with availability score of 0 (from Users_availability): 93
Number of users with at least 10 available months (original dataset): 26
Number of users with fewer than 10 available months (original dataset): 93
The counts match between Users_availability and the original dataset.


Users_Lang_Skills is the name of the dataframe with all the languages and skills valued by merging those datasets.

Users_Lang_Skills_Scored has the scores of the matched skills from the specific job description.

Then added language scores to Users_Lang_Skills_Scored

Then added experience scores to Users_Lang_Skills_Scored. Users do not have more than 2.5 years of experience and data engineering asks for 5. Its not just data engineer, Analyst and Scrum also require 3.5 years which we dont have candidates that fit.


Added availability score to a seperate dataframe because i used Data Analyst instead of engineer since that is the only job with duration. We have an issue where duration is stored in additional infor, therefore to get to duration i ahve to parse through additional info first. If duration is not stored in additional info the code will not extract the duration.

Users_Lang_Skills_Scored base dataframe is Users_Lang_Skills


Since the scoring columns are created after checking the matched skill with the job skill and posting the score value. It means that We are not showing all the skills for candidates. Only the matched ones. Which means our total score is not 136 since not all skills are matched with the job description.

-------------------------------------------------------------------------------------- **Done** ---------------------------------------------------------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

-----------------------------------------------------------

In [ ]:
# Define columns that need to be merged as comma-separated values
merge_columns = {
    "HCK_HEC_LANG": ["LANGUAGE_SKILL_DSC", "LANGUAGE_SKILL_LVL"],
    "HCK_HEC_SKILLS": ["SKILLS_DSC", "DOMAIN_DSC", "LEVEL_VAL"],
    "HCK_HEC_XP": ["MISSION_DSC"]
}

# Aggregate function for merging rows with comma separation
def merge_values(series):
    return ', '.join(series.dropna().astype(str).unique())

# Apply aggregation to each dataset
df_lang = df_lang.groupby("USER_ID").agg({col: merge_values for col in merge_columns["HCK_HEC_LANG"]}).reset_index()
df_skills = df_skills.groupby("USER_ID").agg({col: merge_values for col in merge_columns["HCK_HEC_SKILLS"]}).reset_index()
df_xp = df_xp.groupby("USER_ID").agg({col: merge_values for col in merge_columns["HCK_HEC_XP"]}).reset_index()

# Merge all tables on USER_ID
df_merged = df_user.copy()  # Start with user table (full reference)
df_merged = df_merged.merge(df_lang, on="USER_ID", how="left")
df_merged = df_merged.merge(df_skills, on="USER_ID", how="left")
df_merged = df_merged.merge(df_staffing, on="USER_ID", how="left")
df_merged = df_merged.merge(df_xp, on="USER_ID", how="left")

In [ ]:
df_merged.head()

USER_ID LAST_NAME FIRST_NAME  ANNEES_XP LANGUAGE_SKILL_DSC  \
0  2843838     Ramos      David        1.0    English, French   
1  2479537     Small       Carl        2.5    English, French   
2  2533337     Evans     Carmen        2.0    English, French   
3  2446382   Pittman     Brandi        2.5    French, English   
4  2433124    Thomas      Julie        2.5    French, English   

  LANGUAGE_SKILL_LVL                                         SKILLS_DSC  \
0              100.0              Java, Javascript, PL/SQL, Python, SQL   
1        90.0, 100.0  Oracle Analytics Cloud, Oracle Hyperion Financ...   
2        50.0, 100.0  Anaconda Enterprise, Matlab, GitLab, IBM Cogno...   
3        100.0, 60.0  Microsoft Power BI, Microsoft SQL Server Repor...   
4              100.0                               Snowflake Data Cloud   

                                          DOMAIN_DSC  \
0                               Programming Language   
1  Data Analytics, Financial Close, Financial Pla...   
2  Data Science, Agile Delivery, Financial Planni...   
3  Data Analytics, Data Management, Programming L...   
4                                    Data Management   

                      LEVEL_VAL  MONTH_1  ...  MONTH_4  MONTH_5  MONTH_6  \
0        70.0, 20.0, 50.0, 80.0      100  ...      100      100      100   
1       60.0, 40.0, 100.0, 80.0      100  ...      100      100      100   
2  40.0, 10.0, 20.0, 50.0, 30.0       25  ...       25       25       25   
3        80.0, 50.0, 60.0, 40.0       25  ...       25       25       25   
4                         100.0       25  ...       25       25       25   

   MONTH_7  MONTH_8  MONTH_9  MONTH_10  MONTH_11  MONTH_12  \
0      100      100      100       100       100       100   
1      100      100      100       100       100       100   
2       25       25       25        25        25        25   
3       25       25       50        25        25        25   
4       50       25       25        25        25        25   

                                         MISSION_DSC  
0  Mise en place d’un pipeline RAG pour la migrat...  
1  Mise en œuvre d’une application de simulation ...  
2  Intégration de données, Analyser et résoudre l...  
3  Analyser le système d'information financier, C...  
4  Restitution et transmission des connaissances,...  

[5 rows x 22 columns]

In [ ]:

# Calculate null counts and percentages for each column in df_merged
null_counts = df_merged.isnull().sum()
null_percentages = (null_counts / len(df_merged)) * 100

# Create a DataFrame to display the results
null_analysis = pd.DataFrame({
    'Null Counts': null_counts,
    'Null Percentages': null_percentages
})

# Display the null analysis
null_analysis


Null Counts  Null Percentages
USER_ID                       0          0.000000
LAST_NAME                     0          0.000000
FIRST_NAME                    0          0.000000
ANNEES_XP                     0          0.000000
LANGUAGE_SKILL_DSC            7          5.882353
LANGUAGE_SKILL_LVL            7          5.882353
SKILLS_DSC                   26         21.848739
DOMAIN_DSC                   26         21.848739
LEVEL_VAL                    26         21.848739
MONTH_1                       0          0.000000
MONTH_2                       0          0.000000
MONTH_3                       0          0.000000
MONTH_4                       0          0.000000
MONTH_5                       0          0.000000
MONTH_6                       0          0.000000
MONTH_7                       0          0.000000
MONTH_8                       0          0.000000
MONTH_9                       0          0.000000
MONTH_10                      0          0.000000
MONTH_11                      0          0.000000
MONTH_12                      0          0.000000
MISSION_DSC                   5          4.201681

In [ ]:
# prompt: df merged give me one row with missing values

# Find a row with missing values
row_with_missing_values = df_merged[df_merged.isnull().any(axis=1)].head(1)

row_with_missing_values


USER_ID LAST_NAME FIRST_NAME  ANNEES_XP LANGUAGE_SKILL_DSC  \
5  2817191   Johnson       Carl        1.0    French, English   

  LANGUAGE_SKILL_LVL SKILLS_DSC DOMAIN_DSC LEVEL_VAL  MONTH_1  ...  MONTH_4  \
5              100.0        NaN        NaN       NaN       50  ...       50   

   MONTH_5  MONTH_6  MONTH_7  MONTH_8  MONTH_9  MONTH_10  MONTH_11  MONTH_12  \
5       50       50       50       50       50        50        50        50   

                                         MISSION_DSC  
5  Subject Matter Expert for the implementation a...  

[1 rows x 22 columns]

In [ ]:
# prompt: check if userid 2817191 is in df skills

if 2817191 in df_skills['USER_ID'].values:
  print("User ID 2817191 is present in df_skills.")
else:
  print("User ID 2817191 is not present in df_skills.")


User ID 2817191 is not present in df_skills.


ROWS Are NAN because some user_ids don't have matching rows in other tables like skills

In [ ]:
def split_numeric_column(df, col_name, prefix):
    # Split string values into lists of integers
    # Filter out empty strings before conversion
    df[col_name] = df[col_name].apply(lambda x: [int(float(i)) for i in x.split(", ") if i] if pd.notna(x) else [])

    # Expand into separate columns
    max_length = df[col_name].apply(len).max()  # Find max number of values
    for i in range(max_length):
        df[f"{prefix}_{i+1}"] = df[col_name].apply(lambda x: x[i] if i < len(x) else None)

    # Drop original column after expansion
    df.drop(columns=[col_name], inplace=True)

# Apply function to split `LANGUAGE_SKILL_LVL` and `LEVEL_VAL`
split_numeric_column(df_merged, "LANGUAGE_SKILL_LVL", "LANGUAGE_SKILL_LVL")
split_numeric_column(df_merged, "LEVEL_VAL", "LEVEL_VAL")

In [ ]:
df_merged.head()

USER_ID LAST_NAME FIRST_NAME  ANNEES_XP LANGUAGE_SKILL_DSC  \
0  2843838     Ramos      David        1.0    English, French   
1  2479537     Small       Carl        2.5    English, French   
2  2533337     Evans     Carmen        2.0    English, French   
3  2446382   Pittman     Brandi        2.5    French, English   
4  2433124    Thomas      Julie        2.5    French, English   

                                          SKILLS_DSC  \
0              Java, Javascript, PL/SQL, Python, SQL   
1  Oracle Analytics Cloud, Oracle Hyperion Financ...   
2  Anaconda Enterprise, Matlab, GitLab, IBM Cogno...   
3  Microsoft Power BI, Microsoft SQL Server Repor...   
4                               Snowflake Data Cloud   

                                          DOMAIN_DSC  MONTH_1  MONTH_2  \
0                               Programming Language      100      100   
1  Data Analytics, Financial Close, Financial Pla...      100      100   
2  Data Science, Agile Delivery, Financial Planni...       25       25   
3  Data Analytics, Data Management, Programming L...       25       25   
4                                    Data Management       25        0   

   MONTH_3  ...  LANGUAGE_SKILL_LVL_2  LEVEL_VAL_1  LEVEL_VAL_2  LEVEL_VAL_3  \
0      100  ...                   NaN         70.0         20.0         50.0   
1      100  ...                 100.0         60.0         40.0        100.0   
2       25  ...                 100.0         40.0         10.0         20.0   
3       25  ...                  60.0         80.0         50.0         60.0   
4       25  ...                   NaN        100.0          NaN          NaN   

   LEVEL_VAL_4  LEVEL_VAL_5  LEVEL_VAL_6  LEVEL_VAL_7  LEVEL_VAL_8 LEVEL_VAL_9  
0         80.0          NaN          NaN          NaN          NaN         NaN  
1         80.0          NaN          NaN          NaN          NaN         NaN  
2         50.0         30.0          NaN          NaN          NaN         NaN  
3         40.0          NaN          NaN          NaN          NaN         NaN  
4          NaN          NaN          NaN          NaN          NaN         NaN  

[5 rows x 31 columns]

In [ ]:
def match_language_skill(df):
    english_skills, french_skills = [], []

    for index, row in df.iterrows():
        # Check if LANGUAGE_SKILL_DSC is a string and not null before splitting
        if isinstance(row["LANGUAGE_SKILL_DSC"], str) and not pd.isnull(row["LANGUAGE_SKILL_DSC"]):
            languages = row["LANGUAGE_SKILL_DSC"].split(", ")  # Split into list
        else:
            languages = []  # Assign an empty list if not a string or null

        # Access the split columns directly using the prefix
        # Assuming skill levels are stored in columns like "LANGUAGE_SKILL_LVL_1", "LANGUAGE_SKILL_LVL_2", etc.
        # Get the number of LANGUAGE_SKILL_LVL_ columns
        num_skill_cols = len([col for col in df.columns if col.startswith("LANGUAGE_SKILL_LVL_")])

        skill_levels = [row[f"LANGUAGE_SKILL_LVL_{i+1}"] for i in range(num_skill_cols) if pd.notna(row[f"LANGUAGE_SKILL_LVL_{i+1}"])]
        skill_levels = [int(float(x)) for x in skill_levels]

        # Initialize default values
        english_level = None
        french_level = None

        # Assign skill levels based on language order
        for i, lang in enumerate(languages):
            if lang == "English":
                english_level = skill_levels[i] if i < len(skill_levels) else None  # Handle potential IndexError
            elif lang == "French":
                french_level = skill_levels[i] if i < len(skill_levels) else None  # Handle potential IndexError

        english_skills.append(english_level)
        french_skills.append(french_level)

    df["english_skill_lvl"] = english_skills
    df["french_skill_lvl"] = french_skills
    return df


# Apply function to match language skills
df_merged = match_language_skill(df_merged)

# Drop the original columns if no longer needed
df_merged.drop(columns=["LANGUAGE_SKILL_DSC"] + [col for col in df_merged.columns if col.startswith("LANGUAGE_SKILL_LVL_")], inplace=True)  # drop the split columns


In [ ]:
df_merged.head()

USER_ID LAST_NAME FIRST_NAME  ANNEES_XP  \
0  2843838     Ramos      David        1.0   
1  2479537     Small       Carl        2.5   
2  2533337     Evans     Carmen        2.0   
3  2446382   Pittman     Brandi        2.5   
4  2433124    Thomas      Julie        2.5   

                                          SKILLS_DSC  \
0              Java, Javascript, PL/SQL, Python, SQL   
1  Oracle Analytics Cloud, Oracle Hyperion Financ...   
2  Anaconda Enterprise, Matlab, GitLab, IBM Cogno...   
3  Microsoft Power BI, Microsoft SQL Server Repor...   
4                               Snowflake Data Cloud   

                                          DOMAIN_DSC  MONTH_1  MONTH_2  \
0                               Programming Language      100      100   
1  Data Analytics, Financial Close, Financial Pla...      100      100   
2  Data Science, Agile Delivery, Financial Planni...       25       25   
3  Data Analytics, Data Management, Programming L...       25       25   
4                                    Data Management       25        0   

   MONTH_3  MONTH_4  ...  LEVEL_VAL_2  LEVEL_VAL_3  LEVEL_VAL_4  LEVEL_VAL_5  \
0      100      100  ...         20.0         50.0         80.0          NaN   
1      100      100  ...         40.0        100.0         80.0          NaN   
2       25       25  ...         10.0         20.0         50.0         30.0   
3       25       25  ...         50.0         60.0         40.0          NaN   
4       25       25  ...          NaN          NaN          NaN          NaN   

   LEVEL_VAL_6  LEVEL_VAL_7  LEVEL_VAL_8  LEVEL_VAL_9 english_skill_lvl  \
0          NaN          NaN          NaN          NaN             100.0   
1          NaN          NaN          NaN          NaN              90.0   
2          NaN          NaN          NaN          NaN              50.0   
3          NaN          NaN          NaN          NaN              60.0   
4          NaN          NaN          NaN          NaN               NaN   

   french_skill_lvl  
0               NaN  
1             100.0  
2             100.0  
3             100.0  
4             100.0  

[5 rows x 30 columns]

In [ ]:
# prompt: df merged to csv

# Assuming df_merged is your final DataFrame
df_merged.to_csv('merged_data.csv', index=False)


In [ ]:
# prompt: skills_dsc has comma seperated value process them all like lower trim etc and then find how many unique ones are there and list them

import pandas as pd

# Assuming df_skills is your DataFrame
def process_skills(df_skills):
  """Processes skills data, removes duplicates, and returns unique skills."""

  unique_skills = set()
  if 'SKILLS_DSC' in df_skills.columns:
    for skills in df_skills['SKILLS_DSC'].dropna().astype(str):
      for skill in skills.split(', '):
          processed_skill = skill.strip().lower()
          unique_skills.add(processed_skill)

  return list(unique_skills)

unique_skill_list = process_skills(df_skills)
print("Number of Unique Skills:", len(unique_skill_list))
print("Unique Skills:")
for skill in unique_skill_list:
    print(skill) # Added indentation and print statement

Number of Unique Skills: 134
Unique Skills:
google bigquery
java
microsoft sql server integration services (ssis)
kubernetes
sap bw/4hana
dax
fivetran
amazon quicksight
data galaxy
scala
microsoft azure
qlik view
knime analytics platform
amazon redshift
ibm cognos analytics
project management
neo4j graph database
t-sql
oracle hyperion essbase
teradata database
airbyte
microsoft sql server
ibm cognos tm1
tableau desktop and online
sap financial consolidation
oracle analytics cloud
talend open studio for data quality
prophix conso
atlassian bitbucket
javascript
microsoft power bi
hadoop
ibm netezza performance server
docker
oracle hyperion financial management (hfm)
planning & forecasting
oracle cloud epm
github enterprise
oracle business intelligence enterprise edition
alteryx analytics hub
fluence
dbt
workiva
databricks lakehouse platform
amazon aws cloudformation
sap crystal reports
alteryx designer
pl/sql
sap businessobjects bi
sap financial information management
microsoft fabric
sa

# Translate MISSION_DSC column into english

In [ ]:
pip install deep-translator


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 kB 1.7 MB/s eta 0:00:00


In [ ]:
from deep_translator import GoogleTranslator
import pandas as pd

# Load the dataset
df_merged = pd.read_csv('merged_data.csv')

# Initialize Google Translator
translator = GoogleTranslator(source='auto', target='en')

# Function to translate text safely
def safe_translate(text):
    try:
        if pd.notnull(text) and text.strip():  # Ensure text is not empty
            return translator.translate(text)
    except Exception as e:
        print(f"Translation error for '{text}': {e}")
        return text  # Return original text if translation fails
    return text

# Apply translation
df_merged['mission_dsc_eng'] = df_merged['MISSION_DSC'].astype(str).apply(safe_translate)

# Save the updated dataset
df_merged.to_csv('merged_data_translated.csv', index=False)

print("Translation completed and saved as 'merged_data_translated.csv'")


Translation error for 'Définir et déployer la vision, les objectifs et les plans d’action relatifs à l’Analytique en cohérence avec la stratégie globale de [CLIENT]., Collaborer étroitement avec le [CLIENT] pour stimuler la croissance des revenus et avec le [CLIENT] pour optimiser les opérations., Intervenir en tant que Customer Success Manager sur des comptes stratégiques pour assurer la bonne tenue de nos interventions, maintenir une excellente relation avec le client et générer des opportunités., S’assurer du bon taux d’activité des consultants de la BU et de sa rentabilité, Anticiper les risques et mettre en place des actions préventives et évolutives afin de les gérer, Représenter l’Entreprise sur son périmètre d’activité, faire connaître sa BU et son offre sur le marché., Assumer un rôle d'architecte, superviser les autres architectes et les faire profiter de son expérience pour leur permettre de monter en compétence., En lien avec le Bureau de projet, assurer la bonne tenue des 

Does not work well with very long descriptions, so translation error

In [ ]:
import pandas as pd

# Load both datasets
df_original = pd.read_csv('merged_data.csv')
df_translated = pd.read_csv('merged_data_translated.csv')

# Compare shape (rows, columns)
print(f"Original Dataset: {df_original.shape[0]} rows, {df_original.shape[1]} columns")
print(f"Translated Dataset: {df_translated.shape[0]} rows, {df_translated.shape[1]} columns")

# Check if any rows were lost or columns were added/removed
if df_original.shape == df_translated.shape:
    print("✅ Both datasets have the same number of rows and columns.")
else:
    print("⚠️ There is a mismatch in rows or columns!")

# Check if an extra column 'mission_dsc_eng' was correctly added
if 'mission_dsc_eng' in df_translated.columns:
    print("✅ 'mission_dsc_eng' column is present in the translated dataset.")
else:
    print("❌ 'mission_dsc_eng' column is missing!")


Original Dataset: 119 rows, 30 columns
Translated Dataset: 119 rows, 31 columns
⚠️ There is a mismatch in rows or columns!
✅ 'mission_dsc_eng' column is present in the translated dataset.


In [ ]:
df_translated['mission_dsc_eng'][1]

'Implementation of a budget simulation application (Driver-Based Top-Down Allocation for Rolling Forecast), design and implementation of an application for cost allocation within the framework of the monthly Rolling Forecast process, optimization of closing and provisional processes, collection and validation of needs, planning, realization and monitoring, coordination and management of intervention with the local technical team and financial management'